homeassistant/components/sensor/foobot.py

```diff
@@ -0,0 +1,151 @@
+                'pm': [ATTR_PM2_5, 'µg/m3', 'mdi:cloud'],
+                'tmp': [ATTR_TEMPERATURE, TEMP_CELSIUS, 'mdi:thermometer'],
+                'hum': [ATTR_HUMIDITY, '%', 'mdi:water-percent'],
+                'co2': [ATTR_CARBON_DIOXIDE, 'ppm',
+                        'mdi:periodic-table-co2'],
+                'voc': [ATTR_VOLATILE_ORGANIC_COMPOUNDS, 'ppb',
+                        'mdi:cloud'],
+                'allpollu': [ATTR_FOOBOT_INDEX, '%', 'mdi:percent']}
+
+SCAN_INTERVAL = timedelta(minutes=10)
+PARALLEL_UPDATES = 1
+
+TIMEOUT = 10
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_TOKEN): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+})
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the devices associated with the account."""
+    from foobot_async import FoobotClient
+
+    token = config.get(CONF_TOKEN)
+    username = config.get(CONF_USERNAME)
+
+    client = FoobotClient(token, username,
+                          async_get_clientsession(hass),
+                          timeout=TIMEOUT)
+    data = {}
+    dev = []
+    try:
+        devices = await client.get_devices()
+        _LOGGER.debug("The following devices were found: %s", devices)
+        for device in devices:
+            for sensor_type in SENSOR_TYPES:
+                if sensor_type == 'time':
+                    continue
+                foobot_sensor = FoobotSensor(client, data, device,
+                                             sensor_type)
+                dev.append(foobot_sensor)
+    except (aiohttp.client_exceptions.ClientConnectorError,
+            asyncio.TimeoutError, FoobotClient.TooManyRequests,
+            FoobotClient.InternalError):
+        _LOGGER.exception('Failed to connect to foobot servers.')
+        raise PlatformNotReady
+    except FoobotClient.ClientError:
+        _LOGGER.error('Failed to fetch data from foobot servers.')
+        return
+    async_add_devices(dev, True)
+
+
+class FoobotSensor(Entity):
+    """Implementation of a Foobot sensor."""
+
+    __cache = {}
+    __nb_requests = 0
+
+    def __init__(self, client, data, device, sensor_type):
+        """Initialize the sensor."""
+        self._client = client
+        self._uuid = device['uuid']
+        self._name = 'Foobot {} {}'.format(device['name'],
+                                           SENSOR_TYPES[sensor_type][0])
+        self.type = sensor_type
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return SENSOR_TYPES[self.type][2]
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        try:
+            data = self.__cache[self._uuid][self.type]
```

Comment:
Where is `__cache` defined? Attributes prefixed with `__` are reserved for Python internals and should not be used.

Reply:
This a class attribute so it's adequate according to PEP8:

> __double_leading_underscore: when naming a class attribute, invokes name mangling (inside class FooBar, __boo becomes _FooBar__boo; see below).
> \_\_double_leading_and_trailing_underscore\_\_: "magic" objects or attributes that live in user-controlled namespaces. E.g.\_\_init\_\_, \_\_import\_\_ or \_\_file\_\_. Never invent such names; only use them as documented.

But, point taken. This is causing confusion.

To fix that, I moved the data shared between sensors from a class attribute its own class with one instance shared between all the sensor of one device.

This had the added advantage of being able to use `@Throttle` for fetching the data and I was able to remove all the manual "caching" logic. It's cleaner now.